### PR TITLE
Workflow error documentation

### PR DESCRIPTION
**Pull request by matthew-richerson**
Document how the WLM should use the workflow.status.status and workflow.status.message fields

### CODE REVIEW

File: docs/guides/errors/readme.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+A workflow may fail to complete the desired state for many reasons. The best action for the WLM to take depends on what is preventing the workflow from progressing. DWS communicates error information back to the WLM through a few different fields so the WLM can take the best course of action.
+
+## Drivers and Directives
+
+Most of the work required to move a workflow into the desired state is performed by software outside of DWS. These external pieces of software are called drivers. An example of a driver is the NNF software.
+
+Directives are the #DW lines an end user provides as part of their job. There can be more than one directive in a job, and therefore more than one directive in a DWS workflow.
+
+Directives are processed individually by the underlying drivers. At workflow creation time, the drivers modify the workflow to indicate which workflow states they need to do work for for each of the directives. For each workflow state, DWS then waits for all the drivers to complete their work for each of the directives before marking the entire workflow as complete.
```
Comment (bdevcich) [inline docs/guides/errors/readme.md]: > At workflow creation time, the drivers modify the workflow to indicate which workflow states they need to do work for for each of the directives.

My brain doesn't want to accept the "for for" part even though it makes sense.

Comment (matthew-richerson) [reply docs/guides/errors/readme.md]: I'm all for a different wording. I just couldn't think of anything else

Comment (bdevcich) [reply docs/guides/errors/readme.md]: Here's a start:

> At workflow creation time, the drivers modify the workflow to indicate which workflow states they are responsible for completing. This occurs for each of the directives.

**Review by bdevcich — APPROVED** [review]

File: docs/guides/errors/readme.md (hunk 1 above)
Comment (matthew-richerson) [reply docs/guides/errors/readme.md]: Two sentences. I like it. I'll change it

File: docs/guides/errors/readme.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+```
+
+## WLM Actions
+
+When a workflow fails to progress due to an error, the WLM should change its response based on the type of error. When `workflow.status.status` is `Error`, the workflow will not recover from the error, and the WLM should move the workflow to state `Teardown` immediately. When `workflow.status.status` is `TransientCondition`, the driver code will continue to retry in the background, and the issue may resolve. The WLM should wait a short amount of time before moving the workflow to state `Teardown`. 30 seconds is the recommended wait period.
```
Comment (roehrich-hpe) [inline docs/guides/errors/readme.md]: If you're beginning a sentence with "30", write it out as word.

**Comment by roehrich-hpe** [comment]
I'd like to see another section that dives into some example driver code.  That could even be a separate PR.

I'd llke to see a description of how a driver would clear a TransientCondition, or an Error.  Is the driver required to clear the workflow.status.message?  or the workflow.status.drivers[].message?  (I think I saw that the webhook enforces one of those?)

I'd like to see examples for the driver developer, so they can use the stuff in `resource_error.go` and the stuff at the front of `workflow_types.go` without having to know about, and then rummage in, nnf-sos to find examples.  The driver developer should at least be told where they can find these support pieces.

**Comment by matthew-richerson** [comment]
> I'd like to see another section that dives into some example driver code. That could even be a separate PR.
> 
> I'd llke to see a description of how a driver would clear a TransientCondition, or an Error. Is the driver required to clear the workflow.status.message? or the workflow.status.drivers[].message? (I think I saw that the webhook enforces one of those?)
> 
> I'd like to see examples for the driver developer, so they can use the stuff in `resource_error.go` and the stuff at the front of `workflow_types.go` without having to know about, and then rummage in, nnf-sos to find examples. The driver developer should at least be told where they can find these support pieces.

Okay. I can add that as a separate PR.

**Review by ajfloeder — APPROVED** [review]

**Pull request merged**